js/canvasmanager/events.ts

```diff
@@ -0,0 +1,19 @@
+import { DrawObject } from "./drawobjects/drawobject";
+import { CanvasManager } from "./canvasmanager";
+export interface Event {
+  undo: (cm: CanvasManager) => void;
+}
+export class DrawEvent implements Event {
+  undo(cm: CanvasManager): void {
+    console.assert(cm.objects.length !== 0, "undoing draw with empty objects");
+    cm.objects.pop();
+  }
+}
+export class ClearEvent implements Event {
+  constructor(private objects: DrawObject[]) { }
+
+  undo(cm: CanvasManager): void {
+    console.assert(cm.objects.length === 0, "undoing clear without empty objects");
+    cm.objects = this.objects;
+  }
+}
```

Comment:
Why is this simply not a part of canvasmanager.js? It's not java. There's no need to put everything into many tiny files.

Reply:
There's like 3 events and it felt like it was getting in the way of the main canvas manager code to me 🤷🏻‍♂️

Reply:
canvasmanager.ts would easily be short enough even with the events in it. I'd rather have it in one file. If it **really** needs to be in a separate file, then the directory structure should be e.g. `canvasmanager-events.ts`, `canvasmanager.ts` and `drawobjects/pen.ts`.

Reply:
I think having the events in the same file is kinda noisy, and I don't really see the point of naming something with a prefix when you could just as easily use a directory tbh

But if you want I can merge events.ts and canvasmanager.ts

Reply:
That would be nicer than prefixing or many nested directories